backend/app/models/sample_type.rb

```diff
@@ -0,0 +1,284 @@
+              end
+            end
+          end
+        end
+      end
+    end
+
+    return sample_type_new, false
+
+  end
+
+  # Update existing sample type
+  # - Keeps any existing field types (and changes them as necessary)
+  # - Removes any feild types that no longer exist
+  # - Adds any new feild types
+  # - Also updates allowable field types for each field type
+  # - Any potential errors are handled automatically and silently
+  #
+  # @param st [Hash] the sample type
+  #
+  # @option st[:id] [Int] the id of the sample type
+  # @option st[:name] [String] the name of the sample type
+  # @option st[:description] [String] the description of the sample type
+  # @option st[:field_types] [Hash] the field_type attributes associated with the sample type
+  #
+  # feild_types = {
+  #   id [Int] the id of the existing field type <or> nil if it is new
+  #   name [String] name
+  #   ftype [String]  ftype
+  #   required [String] required - interpreted as Boolen
+  #   array [String] array - interpreted as Boolean
+  #   choices [String] choices
+  #   allowable_field_types [Array] array of allowable field types (for ftype = "sample")
+  # }
+  #
+  # allowable_field_type = {
+  #   id [Int] the id of the existing allowable field type <or> nil if it is new
+  #   sample_type_id [Int] id of the allowable field type
+  # }
+  #
+  # @return the sample type
+  def update(st)
+    input_name = Input.text(st[:name])
+    input_description = Input.text(st[:description])
+
+    # Update name and description if not blank
+    self.name = input_name if input_name
+    self.description = input_description if input_description
+
+    # Update the sample type name and description if valid (otherwise leave as is)
+    self.save if self.valid?
+
+    # List of field type ids
+    # - Used to delete field types that were removed on update
+    # - Initialize with id = 0 to generate the appropriate sql query later
+    field_type_ids = [0]
+
+    # Save field type if there is a field name
+    if st[:field_types].kind_of?(Array)
+      st[:field_types].each do |ft|
+        # List of allowable field type ids for this field_type_id
+        # - Used to delete allowable field types that were removed on update
+        # - Initialize with id = 0 to generate the appropriate sql query later
+        allowable_field_type_ids = [0]
+
+        fid       = Input.number(ft[:id])
+        fname     = Input.text(ft[:name])
+        ftype     = Input.text(ft[:ftype])
+        frequired = Input.boolean(ft[:required])
+        farray    = Input.boolean(ft[:array])
+        fchoices  = Input.text(ft[:choices])
+
+        # Find existing feild_type or create new one
+        sql = "select * from field_types where id = #{fid} and parent_id = #{self.id} limit 1"
+
+        field_type_update = (FieldType.find_by_sql sql)[0] || FieldType.new
+
+        # Set fname if exists and input is not blank
+        # Awkward logic, but that's how it works in v2
```

Comment:
Is this necessary to replicate the front end behavior?

Reply:
Yes, this is necessary to replicate the current UX.

Personally I would do this differently (I have designed and developed similar systems with similar complexity), but it would be a significant change to the UX, would require significant design changes (both front-end and back-end), and would be more complex to implement. 

I'm happy to write up a proposal to review with Mariko and Devin and Sean. I don't think we have the time for that, though.  And I think it is outside of our scope of work (i.e., "v3" defaults to "v2")

Reply:
I think it is OK to deviate from the legacy interface if it is easier and results in a better backend.
(Some of this responsibility may belong in the frontend.)

But as we discussed, go ahead and flag these as a TODO and create an issue with a link to the code.